tensorflow/python/kernel_tests/nn_ops/cudnn_deterministic_base.py

```diff
@@ -71,98 +73,117 @@ def _random_data_op(self, shape):
     in_op = self._random_data_op(in_shape)
     strides = [1, 1, 1, 1]
     padding = 'SAME'
-    dilations = [1, 2, 2, 1]
-    out_op = self._random_out_op(in_shape, filter_shape, strides, padding)
-    filter_gradient_op = nn_ops.conv2d_backprop_filter(
+    dilations = [1, rate, rate, 1]
+    out_op = self._random_out_op(
+        in_shape, filter_shape, strides, padding, dilations)
+    self._assert_reproducible(lambda: nn_ops.conv2d_backprop_filter(
         in_op, filter_shape, out_op, strides=strides, padding=padding,
-        dilations=dilations)
-    self._assert_reproducible(filter_gradient_op)
+        dilations=dilations))
 
+  # A configuration for this test could not be found that exercises
+  # nondeterminism on XLA.
   @test_util.run_cuda_only
-  def testBackwardInputGradient(self):
-    in_shape = LayerShapeNHWC(batch=8, height=32, width=32, channels=8)
-    filter_shape = FilterShape2D(
-        height=7, width=7, in_channels=8, out_channels=128)
-    filter_op = self._random_data_op(filter_shape)
-    strides = [1, 1, 1, 1]
-    padding = 'SAME'
-    out_op = self._random_out_op(in_shape, filter_shape, strides, padding)
-    input_gradient_op = nn_ops.conv2d_backprop_input(
-        in_shape, filter_op, out_op, strides=strides, padding=padding)
-    self._assert_reproducible(input_gradient_op)
+  def testConvBackwardFilterGradientWithDilations(self):
+    self.testConvBackwardFilterGradient(rate=2)
 
   @test_util.run_cuda_only
-  def testBackwardInputGradientWithDilations(self):
-    in_shape = LayerShapeNHWC(batch=8, height=32, width=32, channels=8)
+  def testConvBackwardInputGradient(self, rate=1):
+    in_shape = LayerShapeNHWC(batch=1, height=16, width=16, channels=1)
     filter_shape = FilterShape2D(
-        height=7, width=7, in_channels=8, out_channels=128)
+        height=7, width=7, in_channels=1, out_channels=3)
     filter_op = self._random_data_op(filter_shape)
     strides = [1, 1, 1, 1]
     padding = 'SAME'
-    dilations = [1, 2, 2, 1]
-    out_op = self._random_out_op(in_shape, filter_shape, strides, padding)
-    input_gradient_op = nn_ops.conv2d_backprop_input(
+    dilations = [1, rate, rate, 1]
+    out_op = self._random_out_op(
+        in_shape, filter_shape, strides, padding, dilations)
+    self._assert_reproducible(lambda: nn_ops.conv2d_backprop_input(
         in_shape, filter_op, out_op, strides=strides, padding=padding,
-        dilations=dilations)
-    self._assert_reproducible(input_gradient_op)
+        dilations=dilations))
+
+  # A configuration for this test could not be found that exercises
+  # nondeterminism on XLA.
+  @test_util.run_cuda_only
+  def testConvBackwardInputGradientWithDilations(self):
+    self.testConvBackwardInputGradient(rate=2)
+
+  # A 2D convolution configuration that exercises nondeterminism in the forward
+  # direction (i.e. tf.nn.conv2d) using XLA was not found and therefore no tests
+  # for the backprop of tf.nn.conv2d_transpose have been included.
```

Comment:
For someone to understand this comment, they must know the backprop of conv2d_transpose is equivalent to the forward direction of conv2d. I would just drop this comment, or add the transpose backwards case. It's still useful to test determinism even if the op is unconditionally deterministic, since it's possible the implementation of the op later changes to be nondeterministic.

Reply:
I have removed the comment and added tests for reproducibility of the gradients (wrt both `filters` and `input`) of `tf.nn.conv2d_transpose`. Interestingly, it seems that the backprop to `filters` for `tf.nn.conv2d_transpose` is nondeterministic when determinism is not enabled, presumably because it uses the same functionality as backprop to `filters` for `tf.nn.conv2d`, due to the symmetry.